src/Microsoft.OData.Core/UriParser/Resolver/ODataUriResolver.cs

```diff
@@ -530,5 +539,29 @@ internal static ODataUriResolver GetUriResolver(IServiceProvider container)
                 }
             }
         }
+
+        private static CaseInsensitiveSchemaElementsCache GetCaseInsensitiveSchemaElementsCache(IEdmModel model)
+        {
+            if (model == null)
+            {
+                throw new ArgumentNullException(nameof(model));
+            }
+
+            CaseInsensitiveSchemaElementsCache cache = model.GetAnnotationValue<CaseInsensitiveSchemaElementsCache>(model);
+            if (cache == null)
+            {
+                // There's a chance 2 or more threads can reach here concurrently
+                // for the first N parallel requests, and each will build the cache.
+                // While that is wasteful, it doesn't affect the behaviour of the cache since the model is immutable.
+                // The last cache to be attached to the model will "win" and be used for all subsequent requests.
+                // We can avoid this waste by providing a method that user can call manually to build
+                // the cache before any request is made. But I did not want to add a new method to the public API.
+                // We revisit this if it turns out to be a problem in practice.
+                cache = new CaseInsensitiveSchemaElementsCache(model);

```

Comment:
Why not do this when the model is marked as immutable?

Reply:
I could do that. But that means I'd need to define the cache in the `Microsoft.OData.Edm` and make it public so that it can be accessed from the `Microsoft.OData.Core` lib where the `ODataUriResolver` resides. But maybe that's not as harmful.

Reply:
I'm considering moving this to code to take place when the model is marked as immutable. One downside of that approach compared to the current implementation is that the current implementation only creates the cache if there are actually case-insensitive look ups.

Reply:
That is true.
However, OData services _not in control of **all** clients calling them_ are bound to receive bad requests (or different casing) fairly often, which would trigger creation of the cache.

For Microsoft Graph. this will happen within seconds of start. I do not know enough other OData services to assess whether we're a special case or the norm.

Reply:
@syprieur makes sense. I think if case-insensitive is enabled, the cache should be built even before the first request. But my concern is around building the cache even when case-insensitivity is not enabled. It's beneficial to mark a model as immutable even when you don't support case-insensitive URI resolution. When case-insensitive mode is not enabled, this code path is never activated and therefore the cache is not used (I may have to double-check this).

I'm contemplating introducing a method that triggers the building of the cache, that the user has to call. But I think this is not the ideal developer experience, and it introduces a new method to the public API.

Reply:
Right, my bad! Case-sensitivity is an ODataUriParser setting (via ODataUriResolver), it's not tied to an EdmModel.
My suggestion did not make sense, please ignore this thread.

Reply:
Apparently WebAPI 8 enable case-insensitivity by default. So, it's reasonable to assume most customers have this enabled. This makes me lean more towards your proposal to have it in `MarkAsImmutable()`.

But before that, I'd like to understand whether the current state of the PR would be a significant concern in AGS. The cache is created on the first case-insensitive lookup. However, multiple threads can reach this point concurrently and each try to build the cache independently. In the end, only one will end up committed to the model and the others should be discarded in the next GC. So, if this issue occurs in practice, it will probably occur within the first few seconds of the service running and then disappear (assuming case-insensitive requests are common). Is this one-time transient cost acceptable in AGS?

Reply:
It's fine - if we see latency impact we can always make a warm-up junk request after marking the model as immutable.